### PR TITLE
New feature: marking rooms as unread [MSC2867]

### PR DESCRIPTION
**Pull request by alangecker**
implements https://github.com/matrix-org/matrix-doc/pull/2867
strongly related, but not exactly the same https://github.com/vector-im/element-web/issues/4406
already released within [SchildiChat](https://github.com/SchildiChat/matrix-react-sdk/pull/2)

## Description
This PR...
- ...adds a new button for marking a room as unread
- ...shows a visual indicator based on `com.famedly.marked_unread`
- ...cleares the `com.famedly.marked_unread` when the room gets opened

image
image

## Questions
since I'm still quite unfamiliar with the codebase, I got some questions about this PR:
1. ~~should this react to the stable prefix `m.marked_unread` already as well? or is the usual approach to add it as soon as the MSC is in a further state?~~
2. the current code feels like too much is duplicated but can't really say where to rather put the setting & retrieval of the unread state. maybe even put part of in into https://github.com/matrix-org/matrix-js-sdk? would be happy about any hint :)
3. ~~Any suggestions for an fitting icon?~~

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
Thank you for working on this! Generally, looks great! Just some minor details

<hr></hr>

Could you please include a sign-off as described [here](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.rst#sign-off)?

Also, I think this should probably be hidden under a [labs flag](https://github.com/vector-im/element-web/blob/develop/docs/labs.md) since it is using an unstable prefix

<hr></hr>

> should this react to the stable prefix m.marked_unread already as well? or is the usual approach to add it as soon as the MSC is in a further state?

As far as I know, this usually isn't done.

> the current code feels like too much is duplicated but can't really say where to rather put the setting & retrieval of the unread state. maybe even put part of in into matrix-org/matrix-js-sdk? would be happy about any hint :)

I probably wouldn't put it into the `js-sdk` but you could either create a new file for this or put it into an existing one (e.g. `Unread.ts`)

File: src/i18n/strings/en_EN.json [hunk 1]
```diff
@@ -1038,6 +1038,7 @@
     "Failed to copy": "Failed to copy",
     "Share invite link": "Share invite link",
     "Invite people": "Invite people",
+    "Mark as unread":"Mark as unread",
```
[the remark this answers is not in the document]
Comment (alangecker) [reply src/i18n/strings/en_EN.json]: ah, I was not aware about the script. reverted my change now and executed the script :)

File: src/stores/notifications/RoomNotificationState.ts [hunk 2]
```diff
@@ -110,13 +116,21 @@ export class RoomNotificationState extends NotificationState implements IDestroy
...
-                // We don't have any notified messages, but we might have unread messages. Let's
-                // find out.
-                const hasUnread = Unread.doesRoomHaveUnreadMessages(this.room);
-                if (hasUnread) {
-                    this._color = NotificationColor.Bold;
+                // it is marked as unread?
+                const isMarkedUnread = this.room.getAccountData('com.famedly.marked_unread')?.getContent()?.unread
+
+                if (isMarkedUnread) {
+                    this._color = NotificationColor.Red;
```
[the remark this answers is not in the document]
Comment (alangecker) [reply src/stores/notifications/RoomNotificationState.ts:131]: I thought about the bold as well, but the MSC states:
> When this is true a client should show the room with an indeterminate unread marker. This marker should
be of similar visual importance to a notification badge without the notification count

The Notification badge is usually in red, so I assume, that this means the unread count should also be red?

**Comment by alangecker** [comment]
thanks a lot for the feedback! :)

- I signed off my commits now
- the feature is now behind a feature flag: `feature_mark_unread`
- I hope I resolved all your other points as well :)

File: src/stores/notifications/RoomNotificationState.ts [hunk 3]
```diff
@@ -76,6 +87,10 @@ export class RoomNotificationState extends NotificationState implements IDestroy
         }
     };
 
+    private handleRoomAccountDataUpdate = (ev: MatrixEvent) => {
+        this.updateNotificationState();
```
Comment (SimonBrandner) [inline src/stores/notifications/RoomNotificationState.ts:91]: We should probably only do this if this is the relevant event type

File: src/stores/notifications/RoomNotificationState.ts [hunk 4]
```diff
@@ -24,16 +24,25 @@ import { Room } from "matrix-js-sdk/src/models/room";
 import * as RoomNotifs from '../../RoomNotifs';
 import * as Unread from '../../Unread';
 import { NotificationState } from "./NotificationState";
+import { isRoomMarkedAsUnread } from "../../Rooms";
+import SettingsStore from "../../settings/SettingsStore";
 
 export class RoomNotificationState extends NotificationState implements IDestroyable {
+    settingWatcherRef = null;
```
Comment (SimonBrandner) [inline src/stores/notifications/RoomNotificationState.ts:31]: Could we please give this a less generic name and make it private?

**Review by SimonBrandner — COMMENTED** [review]
Two more tiny things 😀

File: src/stores/notifications/RoomNotificationState.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonBrandner) [reply src/stores/notifications/RoomNotificationState.ts:131]: Yeah, this makes sense 👍

**Comment by alangecker** [comment]
@SimonBrandner
as many as you find! :)

regarding an icon: can I just add any license compatible one or would a design team potentially become unhappy? :D

**Review by SimonBrandner — APPROVED** [review]
LGTM

Regarding the icon: I think you can use any icon that fits the license. Though it's quite probable the design team will suggest their own in the end

**Comment by turt2live** [comment]
@alangecker what's the status of this? It's flagged as WIP and has a bit of a confusing history to it. Is it ready for review by our teams?

**Comment by alangecker** [comment]
@turt2live I still didn't manage to find and integrate an icon, that's why it is still flagged as WIP. From the functionality I wouldn't do any more changes and would be happy about a review :)

also if the design team can imagine deciding for an icon, I would be pleased! :)

File: src/Rooms.ts [hunk 5]
```diff
@@ -143,3 +143,19 @@ function guessDMRoomTargetId(room: Room, myUserId: string): string {
     if (oldestUser === undefined) return myUserId;
     return oldestUser.userId;
 }
+
+export const UNSTABLE_MSC2867_MARKED_UNREAD_TYPE = 'com.famedly.marked_unread';
```
Comment (turt2live) [inline src/Rooms.ts:147]: we should be using the js-sdk's `UnstableValue` class.

File: src/stores/notifications/RoomNotificationState.ts [hunk 6]
```diff
@@ -24,16 +24,25 @@ import { Room } from "matrix-js-sdk/src/models/room";
 import * as RoomNotifs from '../../RoomNotifs';
 import * as Unread from '../../Unread';
 import { NotificationState } from "./NotificationState";
+import { isRoomMarkedAsUnread, UNSTABLE_MSC2867_MARKED_UNREAD_TYPE } from "../../Rooms";
+import SettingsStore from "../../settings/SettingsStore";
 
 export class RoomNotificationState extends NotificationState implements IDestroyable {
+    private featureMarkedUnreadWatcherRef = null;
     constructor(public readonly room: Room) {
```
Comment (turt2live) [inline src/stores/notifications/RoomNotificationState.ts:32]: missing blank line

File: src/stores/notifications/RoomNotificationState.ts [hunk 7]
```diff
@@ -76,6 +87,12 @@ export class RoomNotificationState extends NotificationState implements IDestroy
         }
     };
 
+    private handleRoomAccountDataUpdate = (ev: MatrixEvent) => {
```
Comment (turt2live) [inline src/stores/notifications/RoomNotificationState.ts:90]: this doesn't feel like the right type for account data: there's a ton of missing fields on account data events.

File: src/stores/notifications/RoomNotificationState.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/notifications/RoomNotificationState.ts:133]: A notification badge is grey or red depending on various states, which translates to bold or red.

**Review by turt2live — COMMENTED** [review]
I'm concerned more generally about the MSC itself as this doesn't give a reliable count for the user to rely upon, so they end up trying to discern the difference between manually marked unread, unread, and mentioned. 

I'll copy these comments to the MSC, but overall I don't feel this is a problem that account data should be solving.

**Comment by alangecker** [comment]
seems like there is support for going a different path for that feature (https://github.com/matrix-org/matrix-doc/pull/2867#pullrequestreview-723783823), so I won't continue here for now.

I'm still motivated working on an implementation for this feature, but would need others to lead the spec discussions! :)

**Comment by SpiritCroc** [comment]
@alangecker thank you for this PR! We have included it in SchildiChat-web/desktop: https://github.com/SchildiChat/matrix-react-sdk/pull/2
Do you want to be mentioned in the changelog?

**Comment by alangecker** [comment]
@SpiritCroc I was a bit sad to stop the development, wanted to wait for the spec discussions first, but it's very nice that it's being used now nevertheless! :) 🎉

I would be pleased about credits in changelogs, but I'm also totally fine without! ;)

File: src/Rooms.ts (hunk 5 above)
Comment (alangecker) [reply src/Rooms.ts]: done within https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/6dcd393181a5b605355d306912914fe45eb27050

File: src/stores/notifications/RoomNotificationState.ts (hunk 6 above)
Comment (alangecker) [reply src/stores/notifications/RoomNotificationState.ts:35]: linter is now happy https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/63f5ccbf158bbf152551c406688184f929cf5abb

File: src/stores/notifications/RoomNotificationState.ts (hunk 7 above)
Comment (alangecker) [reply src/stores/notifications/RoomNotificationState.ts:99]: With every occurrence in `matrix-react-sdk` there is `.on("Room.accountData")` used along with `MatrixEvent` as the type for its first argument.

What could be the alternative type? :)

**Comment by alangecker** [comment]
since there is currently no development on the MSC, I thought about pushing this further :)

#### things done
- I've rebased now the branch on top of the current develop and added the additional commits being done by @SpiritCroc  and @su-ex from the [SchildiChat branch](https://github.com/SchildiChat/matrix-react-sdk/pull/2), except keeping this feature still disabled by default.
- resolving most of the open conversations

#### open questions
1. Despite the fact that the MSC is currently still fundamentally in question, can this still be merged behind a feature flag? @turt2live 
2. They commits by @SpiritCroc and @su-ex  are missing a sign-off. Is that an issue and if so how can we resolve that without the complicated path of giving everyone access to this branch and each one has to rebase the commits?^^ couldn't it legally be fine if everyone states here that they sign-off every commit in this PR? :D
3. Two icons from google material design got added https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/e5810fb3edd9d4072d38ce0f2fa4d25b34a1425e. Is that license wise fine? @SimonBrandner

**Comment by SpiritCroc** [comment]
Signed-off-by: Tobias Büttner <dev@spiritcroc.de>

What I know about the license of material design icons I've also written in the commit message of https://github.com/matrix-org/matrix-react-sdk/commit/e5810fb3edd9d4072d38ce0f2fa4d25b34a1425e.

**Comment by su-ex** [comment]
Signed-off-by: Quirin Götz <codeworks@supercable.onl>

**Comment by turt2live** [comment]
So the story on this is that it's in a bit of a questionable bit spec-wise. I think before we commit to maintenance of this feature we'd want to see a slightly more stable spec discussion first. For the time being, I'm going to close this because it's bitrotting beyond reasonable repair, but it would still count as an implementation against the MSC if needed (though there are plenty of other implementations on the MSC as well). The code would still be here for someone to pick up if needed down the line.

Thanks for taking the time to look at it, and apologies for the long turnaround.

**Pull request closed without merging**